pkg/kubelet/cm/cpumanager/policy_static.go

```diff
@@ -326,8 +330,20 @@ func (p *staticPolicy) allocateCPUs(s state.State, numCPUs int, numaAffinity bit
 	result := cpuset.NewCPUSet()
 	if numaAffinity != nil {
 		alignedCPUs := cpuset.NewCPUSet()
-		for _, numaNodeID := range numaAffinity.GetBits() {
-			alignedCPUs = alignedCPUs.Union(allocatableCPUs.Intersection(p.topology.CPUDetails.CPUsInNUMANodes(numaNodeID)))
+
+		numaBits := numaAffinity.GetBits()
+		// If align-by-socket policy option is enabled, NUMA based hint is expanded to
+		// socket aligned hint. It will ensure that first socket aligned available CPUs are
+		// allocated before we try to find CPUs across socket to satisfy allocation request.
+		if p.options.AlignBySocket {
+			socketBits := p.topology.CPUDetails.SocketsInNUMANodes(numaBits...).ToSliceNoSort()
+			for _, socketID := range socketBits {
+				alignedCPUs = alignedCPUs.Union(allocatableCPUs.Intersection(p.topology.CPUDetails.CPUsInSockets(socketID)))
+			}
+		} else {
+			for _, numaNodeID := range numaBits {
+				alignedCPUs = alignedCPUs.Union(allocatableCPUs.Intersection(p.topology.CPUDetails.CPUsInNUMANodes(numaNodeID)))
+			}
 		}
```

Comment:
The way to avoid an if/else and make this a little cleaner here would be to introduce a new function as

```
func (p *staticPolicy) getAlignedCPUs(numaAffinity bitmask.BitMask) {
	alignedCPUs := cpuset.NewCPUSet()
	if p.options.AlignBySocket {
		socketBits := p.topology.CPUDetails.SocketsInNUMANodes(numaAffinity.GetBits()...).ToSliceNoSort()
		for _, socketID := range socketBits {
			alignedCPUs = alignedCPUs.Union(allocatableCPUs.Intersection(p.topology.CPUDetails.CPUsInSockets(socketID)))
		}
		return alignedCPUs
	}

	for _, numaNodeID := range numaBits {
		alignedCPUs = alignedCPUs.Union(allocatableCPUs.Intersection(p.topology.CPUDetails.CPUsInNUMANodes(numaNodeID)))
	}
	
	return alignedCPUs
}
```
And then at the callsite here, just do:
```
alignedCPUs := p.getAlignedCPUs(numaAffinity)
```